src/lib/sw_engine/tvgSwRaster.cpp

```diff
@@ -1361,6 +1361,31 @@ bool rasterClear(SwSurface* surface)
 }
 
 
+void rasterUnpremultiply(SwSurface* surface)
+{
+    //TODO: Create simd avx and neon version
+    for (uint32_t y = 0; y < surface->h; y++) {
+        auto buffer = surface->buffer + surface->stride * y;
+        for (uint32_t x = 0; x < surface->w; ++x) {
+            uint8_t a = buffer[x] >> 24;
+            if (a == 255) {
+                continue;
+            } else if (a == 0) {
+                buffer[x] = 0x00ffffff;
+            } else {
+                uint16_t r = ((buffer[x] >> 8) & 0xff00) / a;
+                uint16_t g = ((buffer[x]) & 0xff00) / a;
+                uint16_t b = ((buffer[x] << 8) & 0xff00) / a;
+                if (r > 0xff) r = 0xff;
+                if (g > 0xff) g = 0xff;
+                if (b > 0xff) b = 0xff;
+                buffer[x] = (a << 24) | (r << 16) | (g << 8) | (b);
```

Comment:
um... I'm not sure if it's appropriate to use  ALPHA_BLEND. But I think it can be expressed like this.
 ALPHA_BLEND(buffer[x], surface->blender.alpha(buffer[x])) 
Or maybe you could just use ALPHA_MULTIPLY for each rgb. What do you think?

Reply:
For unpremultiplication, the color values are divided, not multiplied. I am not sure it can be optimized.